### PR TITLE
PrepareProposal-VoteExtension integration

### PR DESCRIPTION
**Pull request by mconcat**
Please add a description of the changes that this PR introduces and the files that
are the most critical to review.

If this PR fixes an open Issue, please include "Closes #XXX" (where "XXX" is the Issue number) 
so that GitHub will automatically close the Issue when this PR is merged.

### CODE REVIEW

File: state/execution.go
Comment (ValarDragon) [inline state/execution.go:127]: Unsolved question: How should we be handling the block data size. 

from Marko: Idea, leave it up to application to say how many bytes it wants reaped. Then preprocess proposal will reason about what  it wants to keep if it adds data, etc, or removing excess tx data

**Comment by marbar3778** [comment]
@mconcat can you rebase then we can work on reviewing and merging this

File: internal/consensus/byzantine_test.go
Comment (JayT106) [inline internal/consensus/byzantine_test.go:171]: format?

File: internal/consensus/byzantine_test.go
Comment (JayT106) [inline internal/consensus/byzantine_test.go:180]: format?

File: internal/consensus/state.go
Comment (JayT106) [inline internal/consensus/state.go:2062]: do we need the error log in here!

File: internal/consensus/byzantine_test.go
Comment (JayT106) [reply internal/consensus/byzantine_test.go:171]: there are a few lines that need to be formatted. But it's minor.

File: proto/tendermint/abci/types.proto
Comment (marbar3778) [inline proto/tendermint/abci/types.proto:140]: is this the previous blocks votes? how come we are passing this up?

Comment (mconcat) [reply proto/tendermint/abci/types.proto:140]: We need to know about the validator's previous block vote data in order to process related infos inside the PrepareProposal, the main reason is to include the AppDataSelfAuthenticating inside the block header as app_data in case it is needed.

Comment (marbar3778) [reply proto/tendermint/abci/types.proto:140]: could you leave a comment saying this?

**Comment by mconcat** [comment]
One test under light/mbt is failing because of hardcoded json data on BlockID.

**Comment by marbar3778** [comment]
> One test under light/mbt is failing because of hardcoded json data on BlockID.

oh this data comes from informal. @milosevic @ebuchman is there a way we can get updated json data, or is there a way we can do this ourselves?

**Comment by mconcat** [comment]
I am having a nondeterministic error under internal/consensus/state_test.go

**Comment by marbar3778** [comment]
We can merge with failing lightclient/mbt tests.

File: abci/example/kvstore/persistent_kvstore.go
Comment (marbar3778) [inline abci/example/kvstore/persistent_kvstore.go:341]: does this need to be uncommented?

File: proto/tendermint/abci/types.proto
Comment (cmwaters) [inline proto/tendermint/abci/types.proto:141]:
```suggestion
  // votes includes all votes from the previous block. This contains vote extension data that can be used in proposal
  // preparation. The votes here will then form the last commit that gets sent in the proposed block.
```
Does this make more sense? I'm trying to understand this use case better as well as improve the way it is communicated

File: proto/tendermint/types/types.proto
Comment (cmwaters) [inline proto/tendermint/types/types.proto:83]: Few things I want to bring up about this change: 
1) Has this amendment been added to the spec? 
2) What is the justification of including arbitrary application defined bytes in the header? Why can't we just use the app hash to append any extra data the application wants (I don't think there are any size limits on the app hash)
3) What is meant here by non-deterministic? 

I know that having such a field allows for privileged application data to be directly verified via the light client. I'm just uncertain about having application specific data in the tendermint header. I think I'm mostly for such a transition but I would like to gather greater consensus first. Perhaps this is something we can bring up in the tendermint dev call

**Review by cmwaters — APPROVED** [review]
Few high level comments but the implementation looks good to me. I want to make sure such changes are well documented. Do we plan on doing a documentation update after all ABCI++ changes have landed?

**Comment by cmwaters** [comment]
> We can merge with failing lightclient/mbt tests.

It would be nice to move some of the mbt testing code into this repo so we can make such updates. Seeing as the light client has not been touched (and the logic is unlikely to change), it is probably safe to ignore/comment them out

**Comment by marbar3778** [comment]
@mconcat, lets get the questions answered so we can move this forward!

File: abci/example/kvstore/persistent_kvstore.go
Comment (mconcat) [reply abci/example/kvstore/persistent_kvstore.go:341]:
```suggestion
		AppDataSelfAuthenticating: valAddr,
```

**Pull request merged**
